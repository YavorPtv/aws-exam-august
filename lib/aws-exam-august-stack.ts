import * as cdk from 'aws-cdk-lib';
import { LambdaIntegration, RestApi } from 'aws-cdk-lib/aws-apigateway';
import { Runtime, StartingPosition } from 'aws-cdk-lib/aws-lambda';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import * as path from 'node:path';
import { Construct } from 'constructs';
import { AttributeType, BillingMode, StreamViewType, Table } from 'aws-cdk-lib/aws-dynamodb';
import { PolicyStatement, Role, ServicePrincipal } from 'aws-cdk-lib/aws-iam';
import { DynamoEventSource } from 'aws-cdk-lib/aws-lambda-event-sources';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { EmailSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';

export class AwsExamAugustStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const table = new Table(this, 'InvalidEventsTable', {
            partitionKey: {
                name: 'PK',
                type: AttributeType.STRING
            },
            sortKey: {
                name: 'SK',
                type: AttributeType.NUMBER
            },
            timeToLiveAttribute: 'ttl',
            billingMode: BillingMode.PAY_PER_REQUEST,
            stream: StreamViewType.NEW_IMAGE
        });

        const notificationTopic = new Topic(this, 'notificationTopic');

        notificationTopic.addSubscription(
            new EmailSubscription('yavorpetrakiev@gmail.com')
        );

        const schedulerRole = new Role(this, 'SchedulerRole', {
            assumedBy: new ServicePrincipal('scheduler.amazonaws.com'),
        });

        const validateJSONLambdaFunction = new NodejsFunction(this, 'validateJSONLambda', {
            handler: 'handler',
            runtime: Runtime.NODEJS_22_X,
            entry: path.join(__dirname, '../src/validateJSONHandler.ts'),
            environment: {
                TABLE_NAME: table.tableName,
                AWS_ACCOUNT_ID: cdk.Stack.of(this).account,
                SCHEDULER_ROLE_ARN: schedulerRole.roleArn,
                SNS_TOPIC_ARN: notificationTopic.topicArn
            }
        });

        validateJSONLambdaFunction.addToRolePolicy(new PolicyStatement({
            actions: [
                'scheduler:CreateSchedule',
            ],
            resources: ['*']
        }));

        validateJSONLambdaFunction.addToRolePolicy(new PolicyStatement({
            actions: ['iam:PassRole'],
            resources: [schedulerRole.roleArn]
        }));

        table.grantReadWriteData(validateJSONLambdaFunction);

        const deleteHandlerLambdaFunction = new NodejsFunction(this, 'deleteHandler', {
            handler: 'handler',
            runtime: Runtime.NODEJS_22_X,
            entry: path.join(__dirname, '../src/deleteHandler.ts'),
            environment: {
                TABLE_NAME: table.tableName,
                SNS_TOPIC_ARN: notificationTopic.topicArn
            }
        });

        table.grantReadWriteData(deleteHandlerLambdaFunction);
        notificationTopic.grantPublish(deleteHandlerLambdaFunction);

        schedulerRole.addToPolicy(new PolicyStatement({
            actions: ['lambda:InvokeFunction'],
            resources: [deleteHandlerLambdaFunction.functionArn],
        }));

        deleteHandlerLambdaFunction.addEventSource(new DynamoEventSource(table, {
            startingPosition: StartingPosition.LATEST,
            filters: [
                {
                    pattern: JSON.stringify({
                        eventName: ["REMOVE"]
                    })
                }
            ]
        }));

        table.grantReadWriteData(deleteHandlerLambdaFunction);
        notificationTopic.grantPublish(validateJSONLambdaFunction);

        const api = new RestApi(this, 'ValidateJSONApi');
        const resource = api.root.addResource('validate');
        resource.addMethod('POST', new LambdaIntegration(validateJSONLambdaFunction, {
            proxy: true,
        }))
    }
}
