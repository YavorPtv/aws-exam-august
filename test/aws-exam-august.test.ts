import { App } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import { AwsExamAugustStack } from "../lib/aws-exam-august-stack";

test('Stack snapshot test', () => {
    const app = new App();
    const stack = new AwsExamAugustStack(app, 'TestStack');
    const template = Template.fromStack(stack);

    expect(template).toMatchSnapshot();
});
