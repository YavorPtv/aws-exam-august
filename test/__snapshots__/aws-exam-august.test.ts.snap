// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack snapshot test 1`] = `
{
  "Outputs": {
    "ValidateJSONApiEndpoint3D0843D4": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "ValidateJSONApiFFFD6C0E",
            },
            ".execute-api.",
            {
              "Ref": "AWS::Region",
            },
            ".",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "ValidateJSONApiDeploymentStageprod4422656B",
            },
            "/",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "InvalidEventsTable1D2D5E65": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AttributeDefinitions": [
          {
            "AttributeName": "PK",
            "AttributeType": "S",
          },
          {
            "AttributeName": "SK",
            "AttributeType": "N",
          },
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "KeySchema": [
          {
            "AttributeName": "PK",
            "KeyType": "HASH",
          },
          {
            "AttributeName": "SK",
            "KeyType": "RANGE",
          },
        ],
        "StreamSpecification": {
          "StreamViewType": "NEW_IMAGE",
        },
        "TimeToLiveSpecification": {
          "AttributeName": "ttl",
          "Enabled": true,
        },
      },
      "Type": "AWS::DynamoDB::Table",
      "UpdateReplacePolicy": "Retain",
    },
    "SchedulerRole59E73443": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "scheduler.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "SchedulerRoleDefaultPolicy66F774B8": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "deleteHandlerD96F9D76",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "SchedulerRoleDefaultPolicy66F774B8",
        "Roles": [
          {
            "Ref": "SchedulerRole59E73443",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "ValidateJSONApiAccount5CE18BE9": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "ValidateJSONApiFFFD6C0E",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "ValidateJSONApiCloudWatchRole307E86BD",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "ValidateJSONApiCloudWatchRole307E86BD": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "ValidateJSONApiDeployment575DCD952a2f77605a3e14962424c9a3d5516d0c": {
      "DependsOn": [
        "ValidateJSONApiobjectPOST2BF2BE6D",
        "ValidateJSONApiobject2C7CEBA1",
      ],
      "Metadata": {
        "aws:cdk:do-not-refactor": true,
      },
      "Properties": {
        "Description": "Automatically created by the RestApi construct",
        "RestApiId": {
          "Ref": "ValidateJSONApiFFFD6C0E",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "ValidateJSONApiDeploymentStageprod4422656B": {
      "DependsOn": [
        "ValidateJSONApiAccount5CE18BE9",
      ],
      "Properties": {
        "DeploymentId": {
          "Ref": "ValidateJSONApiDeployment575DCD952a2f77605a3e14962424c9a3d5516d0c",
        },
        "RestApiId": {
          "Ref": "ValidateJSONApiFFFD6C0E",
        },
        "StageName": "prod",
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "ValidateJSONApiFFFD6C0E": {
      "Properties": {
        "Name": "ValidateJSONApi",
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "ValidateJSONApiobject2C7CEBA1": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "ValidateJSONApiFFFD6C0E",
            "RootResourceId",
          ],
        },
        "PathPart": "object",
        "RestApiId": {
          "Ref": "ValidateJSONApiFFFD6C0E",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "ValidateJSONApiobjectPOST2BF2BE6D": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "POST",
        "Integration": {
          "IntegrationHttpMethod": "POST",
          "Type": "AWS_PROXY",
          "Uri": {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":apigateway:",
                {
                  "Ref": "AWS::Region",
                },
                ":lambda:path/2015-03-31/functions/",
                {
                  "Fn::GetAtt": [
                    "validateJSONLambdaE910EC0C",
                    "Arn",
                  ],
                },
                "/invocations",
              ],
            ],
          },
        },
        "ResourceId": {
          "Ref": "ValidateJSONApiobject2C7CEBA1",
        },
        "RestApiId": {
          "Ref": "ValidateJSONApiFFFD6C0E",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "ValidateJSONApiobjectPOSTApiPermissionTestStackValidateJSONApiB42F0888POSTobjectFE1700EA": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "validateJSONLambdaE910EC0C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ValidateJSONApiFFFD6C0E",
              },
              "/",
              {
                "Ref": "ValidateJSONApiDeploymentStageprod4422656B",
              },
              "/POST/object",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "ValidateJSONApiobjectPOSTApiPermissionTestTestStackValidateJSONApiB42F0888POSTobject174DABAD": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "validateJSONLambdaE910EC0C",
            "Arn",
          ],
        },
        "Principal": "apigateway.amazonaws.com",
        "SourceArn": {
          "Fn::Join": [
            "",
            [
              "arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":execute-api:",
              {
                "Ref": "AWS::Region",
              },
              ":",
              {
                "Ref": "AWS::AccountId",
              },
              ":",
              {
                "Ref": "ValidateJSONApiFFFD6C0E",
              },
              "/test-invoke-stage/POST/object",
            ],
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "deleteHandlerD96F9D76": {
      "DependsOn": [
        "deleteHandlerServiceRoleDefaultPolicyFA246F0E",
        "deleteHandlerServiceRoleC3F981E4",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "1e039b7165ba4cd2e6848dbc511831dd8f6a2b2b03eab3b665ab09d8853f7862.zip",
        },
        "Environment": {
          "Variables": {
            "SNS_TOPIC_ARN": {
              "Ref": "notificationTopic102BFE2A",
            },
            "TABLE_NAME": {
              "Ref": "InvalidEventsTable1D2D5E65",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "deleteHandlerServiceRoleC3F981E4",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "deleteHandlerDynamoDBEventSourceTestStackInvalidEventsTable495E0A589E92105E": {
      "Properties": {
        "BatchSize": 100,
        "EventSourceArn": {
          "Fn::GetAtt": [
            "InvalidEventsTable1D2D5E65",
            "StreamArn",
          ],
        },
        "FilterCriteria": {
          "Filters": [
            {
              "Pattern": "{"eventName":["REMOVE"]}",
            },
          ],
        },
        "FunctionName": {
          "Ref": "deleteHandlerD96F9D76",
        },
        "StartingPosition": "LATEST",
      },
      "Type": "AWS::Lambda::EventSourceMapping",
    },
    "deleteHandlerServiceRoleC3F981E4": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "deleteHandlerServiceRoleDefaultPolicyFA246F0E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "InvalidEventsTable1D2D5E65",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "notificationTopic102BFE2A",
              },
            },
            {
              "Action": "dynamodb:ListStreams",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": [
                "dynamodb:DescribeStream",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "InvalidEventsTable1D2D5E65",
                  "StreamArn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "deleteHandlerServiceRoleDefaultPolicyFA246F0E",
        "Roles": [
          {
            "Ref": "deleteHandlerServiceRoleC3F981E4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "notificationTopic102BFE2A": {
      "Type": "AWS::SNS::Topic",
    },
    "notificationTopicyavorpetrakievgmailcomB2331866": {
      "Properties": {
        "Endpoint": "yavorpetrakiev@gmail.com",
        "Protocol": "email",
        "TopicArn": {
          "Ref": "notificationTopic102BFE2A",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "validateJSONLambdaE910EC0C": {
      "DependsOn": [
        "validateJSONLambdaServiceRoleDefaultPolicyC0966E16",
        "validateJSONLambdaServiceRoleCC49DD3E",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "58e80a88ec16d2c206857f896ab8da1b94a80c8555b2699799195f812037758a.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_ACCOUNT_ID": {
              "Ref": "AWS::AccountId",
            },
            "SCHEDULER_ROLE_ARN": {
              "Fn::GetAtt": [
                "SchedulerRole59E73443",
                "Arn",
              ],
            },
            "SNS_TOPIC_ARN": {
              "Ref": "notificationTopic102BFE2A",
            },
            "TABLE_NAME": {
              "Ref": "InvalidEventsTable1D2D5E65",
            },
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "validateJSONLambdaServiceRoleCC49DD3E",
            "Arn",
          ],
        },
        "Runtime": "nodejs22.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "validateJSONLambdaServiceRoleCC49DD3E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "validateJSONLambdaServiceRoleDefaultPolicyC0966E16": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "scheduler:CreateSchedule",
              "Effect": "Allow",
              "Resource": "*",
            },
            {
              "Action": "iam:PassRole",
              "Effect": "Allow",
              "Resource": {
                "Fn::GetAtt": [
                  "SchedulerRole59E73443",
                  "Arn",
                ],
              },
            },
            {
              "Action": [
                "dynamodb:BatchGetItem",
                "dynamodb:GetRecords",
                "dynamodb:GetShardIterator",
                "dynamodb:Query",
                "dynamodb:GetItem",
                "dynamodb:Scan",
                "dynamodb:ConditionCheckItem",
                "dynamodb:BatchWriteItem",
                "dynamodb:PutItem",
                "dynamodb:UpdateItem",
                "dynamodb:DeleteItem",
                "dynamodb:DescribeTable",
              ],
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "InvalidEventsTable1D2D5E65",
                    "Arn",
                  ],
                },
                {
                  "Ref": "AWS::NoValue",
                },
              ],
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "notificationTopic102BFE2A",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "validateJSONLambdaServiceRoleDefaultPolicyC0966E16",
        "Roles": [
          {
            "Ref": "validateJSONLambdaServiceRoleCC49DD3E",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
